hydromt_wflow/workflows/gauges.py

```diff
@@ -61,13 +64,33 @@ def gauge_map_uparea(
     if uparea_name not in ds:
         raise ValueError(f"uparea_name {uparea_name} not found in ds.")
 
+    # Find if there are any nodata in gdf["uparea"]
+    if gdf["uparea"].isna().any():
+        logger.warning(
+            "There are NaN values in the gauges uparea."
+            "Uparea from wflow will be used ie no snapping if mask is None."
+        )
+        uparea_wflow = ds[uparea_name].raster.sample(gdf, wdw=0)
+        # Replace nan values in gdf with uparea from wflow
+        gdf["uparea"] = gdf["uparea"].fillna(uparea_wflow.to_pandas())
+
+    ds = ds.copy()
+    # Add mask to ds
+    if mask is not None:
+        ds["mask"] = xr.DataArray(mask, dims=(ds.raster.y_dim, ds.raster.x_dim))
+
     ds_wdw = ds.raster.sample(gdf, wdw=wdw)
+    # Mask valid cells in ds_wdw
+    if mask is not None:
+        ds_wdw[uparea_name] = ds_wdw[uparea_name].where(
+            ds_wdw["mask"], ds_wdw[uparea_name].raster.nodata
+        )
     logger.debug(
         f"Snapping gauges points to best matching uparea cell within wdw (size={wdw})."
     )
     upa0 = xr.DataArray(gdf["uparea"], dims=("index"))
     upa_dff = np.abs(ds_wdw[uparea_name].where(ds_wdw[uparea_name] > 0).load() - upa0)
-    upa_check = np.logical_or((upa_dff / upa0) <= rel_error, upa_dff <= abs_error)
+    upa_check = np.logical_and((upa_dff / upa0) <= rel_error, upa_dff <= abs_error)
     # find best matching uparea cell in window
     i_wdw = upa_dff.argmin("wdw").load()
 
```

Comment:
I have a test case where one gauges does not fall within the range where there is a search for a river. This causes all `NaN` values in 'upa_dff' for a certain index. This causes a crash when trying to apply `argmin` to 'upa_dff'.

We should I guess account for gauges that are unrealistic (in terms of where the rivers are) and out of reach of the rivers.

Reply:
There was indeed a bug here that should be now fixed! Your example still does not work because I added a check if all gauges have uparea==NaN as users should either use no snapping or in your case the default snap_to_river method rather than snap_uparea.